### PR TITLE
etcd-renew - Add extra validation and remove IsLessThanMinDuration

### PR DESCRIPTION
**Pull request by SrinivasAtmakuri**
Added extra validation to verify if etcd certificates are renewed Also, removed isLessThanMinDuration code block so we can even renew clusters whose expiry is more than 6 months duration

### Which issue this PR addresses:

Fixes

### CODE REVIEW

**Comment by SrinivasAtmakuri** [comment]
Currently testing these changes using a test cluster.

**Review by petrkotas — APPROVED** [review]
Nice simplification @srinivasAtmakuri

**Review by petrkotas — DISMISSED** [review]
Nice simplification @srinivasAtmakuri

File: pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go [hunk 1]
```diff
@@ -299,6 +300,44 @@ func (e *etcdrenew) validateEtcdCertsExistsAndExpiry(ctx context.Context) error
 	return nil
 }
 
+func (e *etcdrenew) validateEtcdCertsRenewed(ctx context.Context) error {
```
Comment (s-amann) [inline pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go:303]: I don't see any unit tests for this new method.

**Review by s-amann — DISMISSED** [review]
can we add tests to ensure `validateEtcdCertsRenewed` works when 1, all, and none of the certs are renewed?

File: pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go [hunk 2]
```diff
@@ -299,6 +303,47 @@ func (e *etcdrenew) validateEtcdCertsExistsAndExpiry(ctx context.Context) error
...
+		// etcd operator renews certificates for another 3 years, 1000+ days (3*365)
+		e.log.Infof("certificate '%s' expiration date is '%s'", secretname, certData[0].NotAfter)
+		if utilcert.DaysUntilExpiration(certData[0]) < 1000 {
+			isError = true
+			e.log.Errorf("certificate %s is not renewed successfully.", secretname)
+		}
+	}
+
+	if isError {
+		return api.NewCloudError(http.StatusInternalServerError, api.CloudErrorCodeInternalServerError, "", "etcd certificates renewal not successful, as atleast one or all certificates are not renewed")
```
Comment (s-amann) [inline pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go:340]:
```suggestion
		return api.NewCloudError(http.StatusInternalServerError, api.CloudErrorCodeInternalServerError, "", "etcd certificates renewal not successful, as at least one or all certificates are not renewed")
```

**Review by kimorris27 — COMMENTED** [review]
Nice work! I agree with @s-amann's request for unit tests for the new method but don't have any other comments.

File: pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go (hunk 1 above)
Comment (SrinivasAtmakuri) [reply pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go:306]: Added a new test case. Thanks.

File: pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go (hunk 2 above)
Comment (SrinivasAtmakuri) [reply pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go]: Made the change

File: pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go (hunk 1 above)
Comment (kimorris27) [reply pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go:306]: The added test cases look good, but I was expecting a new test function for `validateEtcdCertsRenewed` (i.e., `TestValidateEtcdCertsRenewed`, to _just_ test this single method). Are you open to adding that as well?

File: pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go [hunk 3]
```diff
@@ -299,6 +303,47 @@ func (e *etcdrenew) validateEtcdCertsExistsAndExpiry(ctx context.Context) error
 	return nil
 }
 
+func (e *etcdrenew) validateEtcdCertsRenewed(ctx context.Context) error {
```
Comment (AldoFusterTurpin) [inline pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go:306]: I see that there is a commit for adding a new test to pkg/frontend/admin_openshiftcluster_etcdcertificaterenew_test.go
but I think it would be a better idea to add a unit test for this function.

The problem is that we are testing the functionality of the business logic (certs validation) creating the whole infra structs and mocks when it could be simplified a bit.

From a testing perspective, in validateEtcdCertsRenewed(...) you could first take all the secrets, store them in an slice, or in a map (from secretName to secret value) and then iterate over the map and do the whole validation. Doing that, we make the function more "pure" avoiding mocking e.k and we can just focus on the inputs and outputs and not side effects. We can also, if we want, get a counter of how many certs were not correctly renewed or even the names of those. The important part is making the function more testable without the mock overhead, in the below example I use the counter variant but we can use whatever we want (like names using a slice of strings for the invalid ones).

something like
```go
//definition of your original function
func (e *etcdrenew) validateEtcdCertsExistsAndExpiry(ctx context.Context) error { 
	e.log.Infoln("validating if etcd certs exists and expiry")

	for _, secretname := range e.secretNames {
		e.log.Infof("validating secret %s", secretname)
		cert, err := e.k.KubeGet(ctx, "Secret", namespaceEtcds, secretname)
		if err != nil {
			return err
		}

		var u unstructured.Unstructured
		var secret corev1.Secret
		if err = json.Unmarshal(cert, &u); err != nil {
			return err
		}
		err = kruntime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &secret)
		if err != nil {
			return err
		}
		_, certData, err := utilpem.Parse(secret.Data[corev1.TLSCertKey])
		if err != nil {
			return err
		}
		if utilcert.IsCertExpired(certData[0]) {
			return api.NewCloudError(http.StatusInternalServerError, api.CloudErrorCodeInternalServerError, "", "secret %s is already expired, quitting.", secretname)
		}
	}
	e.log.Infoln("Etcd certs exits, are not expired")

	return nil
}

// this is the pure function we really want to test without the mocks
func (e *etcdrenew) validateEtcdCertsRenewed(ctx context.Context) error { 
	e.log.Infoln("validating if etcd certs are renewed")

	certs := make(map[string][]byte)
	for _, secretname := range e.secretNames {
		e.log.Infof("getting secret %s", secretname)
		cert, err := e.k.KubeGet(ctx, "Secret", namespaceEtcds, secretname)
		if err != nil {
			return err
		}
		certs[secretname] = cert
	}

	notRenewedCerts, err := getNotRenewedCerts(e.log, certs)
	if err != nil {
		return err
	}

	if notRenewedCerts > 0 {
		errorMsg := fmt.Sprintf("%v certificates were not renewed", notRenewedCerts)
		return api.NewCloudError(http.StatusInternalServerError, api.CloudErrorCodeInternalServerError, "", errorMsg)
	}

	e.log.Infoln("etcd certificates are successfully renewed")
	return nil
}
```
You will need "encodeSecret" from ~/git/ARO-RP/pkg/frontend/admin_openshiftcluster_etcdcertificaterenew_test.go to test the previous function.

So the main idea is to try to avoid testing functionality in code that includes mocks and just use the mocks to ensure that given X I get Y, but NOT given X I get Y and the functionality should be Z, that last thing should be tested independently from the mocks if possible. 

I think in that way our codebase would be much healthier as we would need less mocks to test business logic.
🙂

**Review by AldoFusterTurpin — CHANGES_REQUESTED** [review]
Thank you for this! I just left some optional suggestions and one important comment about testability, in case you find it useful.

Thanks! 
🙂

File: pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go [hunk 4]
```diff
@@ -287,9 +294,6 @@ func (e *etcdrenew) validateEtcdCertsExistsAndExpiry(ctx context.Context) error
 		if err != nil {
 			return err
 		}
-		if !utilcert.IsLessThanMinimumDuration(certData[0], utilcert.DefaultMinDurationPercent) && e.mode != "renewed" {
-			return api.NewCloudError(http.StatusInternalServerError, api.CloudErrorCodeInternalServerError, "", "secret %s is not near expiry, quitting.", secretname)
-		}
```
Comment (jaitaiwan) [inline pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go:292]: Would be good to understand why this was removed?

File: pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go [hunk 5]
```diff
@@ -299,6 +303,47 @@ func (e *etcdrenew) validateEtcdCertsExistsAndExpiry(ctx context.Context) error
 	return nil
 }
 
+func (e *etcdrenew) validateEtcdCertsRenewed(ctx context.Context) error {
+	e.log.Infoln("validating if etcd certs are renewed")
+	isError := false
+
+	for _, secretname := range e.secretNames {
+		e.log.Infof("validating secret %s", secretname)
+		cert, err := e.k.KubeGet(ctx, "Secret", namespaceEtcds, secretname)
```
Comment (jaitaiwan) [inline pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go:312]: Perhaps a retry on this would be good?

File: pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go [hunk 6]
```diff
@@ -299,6 +303,47 @@ func (e *etcdrenew) validateEtcdCertsExistsAndExpiry(ctx context.Context) error
...
+		if err != nil {
+			return err
+		}
+		_, certData, err := utilpem.Parse(secret.Data[corev1.TLSCertKey])
+		if err != nil {
+			return err
+		}
+
+		// etcd operator renews certificates for another 3 years, 1000+ days (3*365)
+		e.log.Infof("certificate '%s' expiration date is '%s'", secretname, certData[0].NotAfter)
```
Comment (jaitaiwan) [inline pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go:332]: `certData[0]` is a potential index out of range error. We're not testing to make sure there is at least 1 item in the certData array.

**Review by jaitaiwan — DISMISSED** [review]
There's a single place where there's a potential for an index out of range panic to happen. Once that's resolved I vote we merge and immediately do another PR to encompass the feedback received here. I don't think any of the feedback is a merge blocker for getting the functionality in the release.

**Comment by jaitaiwan** [comment]
I was testing out the push feature before running tests locally to see if my code had worked, but I've added the catch for the potential index panic, so if the CI passes this PR is ready to merge.

**Review by yjst2012 — APPROVED** [review]
Same as @hlipsig with the believe of this brings more benefits, I'm approving this PR now, @SrinivasAtmakuri pls address the suggestions in a follow up session, thank you.
